torch/distributions/kl.py

```diff
@@ -128,13 +128,13 @@ def _x_log_x(tensor):
     return tensor * tensor.log()
 
 
-def _batch_trace(bmat):
+def _batch_trace_XXT(bmat):
     """
-    Utility function for calculating the trace of matrices with arbitrary trailing batch dimensions
+    Utility function for calculating the trace of XX^{T} with X having arbitrary trailing batch dimensions
     """
     mat_size = bmat.size(-1)
-    values = torch.stack([torch.trace(M) for M in bmat.reshape((-1, mat_size, mat_size))])
-    return values.view(bmat.shape[:-2])
+    flat_trace = bmat.reshape(-1, mat_size * mat_size).pow(2).sum(-1)
```

Comment:
:+1: this looks better, as equivalent to `.pow(2).sum(-1).sum(-1)`

Reply:
@fehiepsi thanks for pointing out this simplification. I had to work out the math, and see you were almost right, just needed one more `.sum(-1)` I believe.